src/tools/snapshot/nodes/PromiseSerializationNodes.java

```diff
@@ -0,0 +1,202 @@
+package tools.snapshot.nodes;
+
+import java.util.List;
+
+import com.oracle.truffle.api.dsl.GenerateNodeFactory;
+import com.oracle.truffle.api.dsl.Specialization;
+
+import som.interpreter.Types;
+import som.interpreter.actors.EventualMessage;
+import som.interpreter.actors.EventualMessage.PromiseMessage;
+import som.interpreter.actors.SPromise;
+import som.interpreter.actors.SPromise.Resolution;
+import som.interpreter.actors.SPromise.SResolver;
+import som.interpreter.objectstorage.ClassFactory;
+import tools.snapshot.SnapshotBuffer;
+import tools.snapshot.deserialization.DeserializationBuffer;
+import tools.snapshot.deserialization.FixupInformation;
+
+
+public abstract class PromiseSerializationNodes {
+
+  @GenerateNodeFactory
+  public abstract static class PromiseSerializationNode extends AbstractSerializationNode {
+
+    public PromiseSerializationNode(final ClassFactory classFact) {
+      super(classFact);
+    }
+
+    @Specialization(guards = "prom.isCompleted()")
+    public void doResolved(final SPromise prom, final SnapshotBuffer sb) {
+      int base = sb.addObject(prom, classFact, 1 + Long.BYTES);
+
+      // resolutionstate
+      switch (prom.getResolutionStateUnsync()) {
+        case SUCCESSFUL:
+          sb.putByteAt(base, (byte) 1);
+          break;
+        case ERRONEOUS:
+          sb.putByteAt(base, (byte) 2);
+          break;
+        default:
+          throw new IllegalArgumentException("This shoud be unreachable");
+      }
+
+      Object value = prom.getValue();
+      if (!sb.containsObject(value)) {
+        Types.getClassOf(value).serialize(value, sb);
+      }
+      sb.putLongAt(base + 1, sb.getObjectPointer(value));
+    }
+
+    @Specialization(guards = "!prom.isCompleted()")
+    public void doUnresolved(final SPromise prom, final SnapshotBuffer sb) {
```

Comment:
Again, a looong method.
Can't you break the steps into smaller methods with nice names?